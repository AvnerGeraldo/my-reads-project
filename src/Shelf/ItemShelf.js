import React from 'react'
import PropTypes from 'prop-types'

//Assets
import './ItemShelf.css'

const ItemShelf = ({ bookData, changeBookShelf }) => {
    //Vars
    const bookTitle = bookData.title !== undefined ? bookData.title : ''
    const backgroundImage = bookData.imageLinks !== undefined && bookData.imageLinks.thumbnail !== undefined ? `url(${bookData.imageLinks.thumbnail}}` : null
    const valueSelected = (bookData.shelf === null || bookData.shelf === undefined ? 'none' : bookData.shelf)
    const listAuthors = bookData.authors !== undefined ? bookData.authors : []

    //Style
    const styleBookCover = {
        backgroundImage,
        width: 128,
        height: 193
    }    

    return (
        <div className="book">
            <div className="book-top">
            <div className="book-cover" style={styleBookCover}></div>
                <div className="book-shelf-changer">
                    <select value={valueSelected} onChange={(e) => changeBookShelf(bookData, e.target.value)}>
                        <option value="move" disabled>Move to...</option>
                        <option value="currentlyReading">Currently Reading</option>
                        <option value="wantToRead">Want to Read</option>
                        <option value="read">Read</option>
                        <option value="none">None</option>
                    </select>
                </div>
            </div>
            <div className="book-title">{bookTitle}</div>
            <div className="book-authors">{listAuthors.join(' , ')}</div>
        </div>
    )
}

const { object, func } = PropTypes
ItemShelf.propTypes = {
    bookData: object.isRequired,
    changeBookShelf: func.isRequired
}

export default ItemShelf