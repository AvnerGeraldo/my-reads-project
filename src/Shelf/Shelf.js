import React from 'react'
import propTypes from 'prop-types'

//Components
import ItemShelf from './ItemShelf'

const Shelf = ({ data, chooseShelf, changeBookShelf }) => {
    let shelf;

    switch (chooseShelf) {
        case 'Currently Reading':
            shelf = 'currentlyReading'
            break;
        case 'Want to Read':
            shelf = 'wantToRead'
            break;
        case 'Read':
            shelf = 'read'
            break;
        default:
            shelf = null
    }

    const listShelfData = data.filter((book) => {
        if (shelf === null) {
            return book
        }

        return book.shelf === shelf
    })

    const styleRow = {
        padding: "0px 10px",
        minWidth: "675px",
        margin: "0 auto",
        maxWidth: "800px"
    }
    
    return (
        <div className="row" style={styleRow}>
            {(listShelfData.length === 0) ? 
            (<div className="no-results">Não há dados para serem exibidos!</div>) :
                listShelfData.map((item, index) => ( 
                <ItemShelf 
                    key={index} 
                    bookData={item} 
                    changeBookShelf={changeBookShelf}/> 
            ))}
        </div>
    )
}

const { func, string, array } = propTypes
Shelf.propTypes = {
    data: array.isRequired,
    chooseShelf: string.isRequired,
    changeBookShelf: func.isRequired
}

export default Shelf