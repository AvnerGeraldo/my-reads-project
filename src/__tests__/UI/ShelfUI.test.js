import React from 'react'
import { shallow, mount } from 'enzyme'

//Component
import Shelf from '../../Shelf/Shelf'

const props = {
    data: [
        {
            title: 'book title 1',
            imageLinks: {
                thumbnail: 'image-link 1'
            },
            shelf: 'wantToRead',
            authors: ['author1', 'author2']
        },
        {
            title: 'book title 2',
            imageLinks: {
                thumbnail: 'image-link 2'
            },
            shelf: 'read',
            authors: ['author1']
        },
        {
            title: 'book title 3',
            imageLinks: {
                thumbnail: 'image-link 3'
            },
            shelf: 'wantToRead',
            authors: []
        }
    ], 
    chooseShelf: 'currentlyReading', 
    changeBookShelf: jest.fn()
}

describe('[Component] Shelf', () => {
    it('render correctly a Shelf', () => {
        const wrapper = shallow(<Shelf {...props}/>)
        
        expect(wrapper).toMatchSnapshot()
    })

    it('rendered a message error when data shelf is empty', () => {
        const newProps = { ...props, data: [] }        
        const wrapper = mount(<Shelf {...newProps}/>)
        
        expect(wrapper.find('.no-results')).toHaveLength(1)
    })

    it('maps ItemsShelf with the same length passed throught props', () => {
        const wrapper = mount(<Shelf {...props}/>)

        expect(wrapper.find('.book')).toHaveLength(props.data.length)
    })
})