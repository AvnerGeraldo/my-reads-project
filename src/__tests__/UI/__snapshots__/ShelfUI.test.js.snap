// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Shelf render correctly a Shelf 1`] = `
<div
  className="row"
  style={
    Object {
      "margin": "0 auto",
      "maxWidth": "800px",
      "minWidth": "675px",
      "padding": "0px 10px",
    }
  }
>
  <ItemShelf
    bookData={
      Object {
        "authors": Array [
          "author1",
          "author2",
        ],
        "imageLinks": Object {
          "thumbnail": "image-link 1",
        },
        "shelf": "wantToRead",
        "title": "book title 1",
      }
    }
    changeBookShelf={[Function]}
    key="0"
  />
  <ItemShelf
    bookData={
      Object {
        "authors": Array [
          "author1",
        ],
        "imageLinks": Object {
          "thumbnail": "image-link 2",
        },
        "shelf": "read",
        "title": "book title 2",
      }
    }
    changeBookShelf={[Function]}
    key="1"
  />
  <ItemShelf
    bookData={
      Object {
        "authors": Array [],
        "imageLinks": Object {
          "thumbnail": "image-link 3",
        },
        "shelf": "wantToRead",
        "title": "book title 3",
      }
    }
    changeBookShelf={[Function]}
    key="2"
  />
</div>
`;
